crates/agent/src/directives/storage_mappings.rs

```diff
@@ -0,0 +1,294 @@
+use std::io::Write;
+
+use crate::{directives::JobStatus, logs};
+use agent_sql::directives::{
+    storage_mappings::{
+        fetch_storage_mappings, upsert_storage_mapping, user_has_admin_capability, StorageMapping,
+    },
+    Row,
+};
+use anyhow::Context;
+use serde::{Deserialize, Serialize};
+use sqlx::types::Uuid;
+use validator::Validate;
+
+#[derive(Debug, Deserialize, Serialize, Validate, schemars::JsonSchema)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Directive {}
+
+#[derive(Debug, Deserialize, Serialize, Validate, schemars::JsonSchema)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Claims {
+    add_store: models::Store,
+    catalog_prefix: models::Prefix,
+}
+
+#[tracing::instrument(skip_all, ret, err, fields(row.user_claims))]
+pub async fn apply(
+    _: Directive,
+    row: Row,
+    logs_tx: &logs::Tx,
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+) -> anyhow::Result<JobStatus> {
+    let detail = format!(
+        "updated by user {} via applied directive {}",
+        row.user_id, row.apply_id
+    );
+    let (collection_data, recovery) = match validate(txn, logs_tx, row).await {
+        Ok(c) => c,
+        Err(err) => {
+            return Ok(JobStatus::invalid_claims(err));
+        }
+    };
+
+    let ProposedMapping {
+        catalog_prefix,
+        spec,
+    } = collection_data;
+    upsert_storage_mapping(&detail, &catalog_prefix, spec, txn).await?;
+    let ProposedMapping {
+        catalog_prefix,
+        spec,
+    } = recovery;
+    upsert_storage_mapping(&detail, &catalog_prefix, spec, txn).await?;
+
+    Ok(JobStatus::Success)
+}
+
+pub struct ProposedMapping {
+    catalog_prefix: String,
+    spec: models::StorageDef,
+}
+
+fn add_store(stores: &mut models::StorageDef, store: models::Store) {
+    // If there's already an equivalent store, then remove it so that we don't end up with
+    // duplicates. This could happen if someone added store A, then store B, then store A again.
+    stores.stores.retain(|s| s != &store);
+    stores.stores.insert(0, store);
+}
+
+async fn validate(
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    logs_tx: &logs::Tx,
+    row: Row,
+) -> anyhow::Result<(ProposedMapping, ProposedMapping)> {
+    let claims: Claims =
+        serde_json::from_str(row.user_claims.get()).context("parsing user_claims")?;
+
+    if !claims.catalog_prefix.ends_with('/') {
+        anyhow::bail!("invalid prefix, must end with '/'");
+    }
+
+    // Storage mappings can only be updated for an entire tenant. We may one day wish to support
+    // updates to narrower prefixes, but are trying to keep it simple for now.
+    if claims.catalog_prefix.matches('/').count() > 1 {
+        anyhow::bail!("catalog prefix contains too many path segments. Only top-level tenant prefixes can have storage mappings altered");
+    }
+
+    // Note: we must assert that user has admin capability for the _entire tenant_, even if in the
```

Comment:
This constraint is fine for now, but I suspect it's too strict and that it is desired to let a user who admins `a/b/` to be able to update the storage mapping for that prefix.